home/rvagg/2021-02-exploration-report-small-filecoin-deals.md

```diff
@@ -0,0 +1,134 @@
+ * `lotus client`
+   * `lotus client import <path>` to import file(s) into the client (somewhere?), presumably as UnixFS (?), gets a CID
+   * `lotus client local` shows you what you have imported and their CIDs
+   * `lotus client balances` shows you funds that you have active in the "market" (where you have to have funds in order to make deals etc. i.e. out of your wallet)
+   * `lotus client find <CID>` lets you query the network for a CID
+
+Deals:
+
+   * `lotus client query-ask <minerID>` ask a miner for their deal terms - returns FIL asking price per GiB per epoch (30 seconds) for both regular and verified deals, along with minimum and maximum deal sizes.
+   * `lotus client list-asks` will do a scan of the network for miners with power and do a `query-ask` on each of them, reporting the output to stdout, one miner per line (~450)
+   * `lotus client deal` without additional arguments takes you through a prompted deal-making process that asks for your data CID, a miner ID, whether it's a verified deal; it queries the miner, asks you to confirm the asking price and starts the deal process if you say `yes`.
+   * `lotus client list-deals` to show you deals in progress and their status (Note: the status identifiers are not self-explainatory, the best source of documentation seems to be the markets code: https://github.com/filecoin-project/go-fil-markets/blob/master/storagemarket/dealstatus.go)
+   * `lotus client deal-stats` seems to be the best way to view the outcome of your deals once they disappear off `list-deals`
+   * The Lotus daemon log also has output related to deals, you can grep for the status, e.g. `grep StorageDealError ~/var/lotus-daemon.log` (mostly just `"error reading Response message: EOF"`)
+
+### Making deals
+
+* Imported a 9.6 Mb (`10016091` b) JPG file to Lotus with `lotus client import`. Some notes about sizes:
+  * Note that data is _padded_ when making deals. First 2 empty bits per 254 bytes are added (called "fr32 padding", `Math.ceil(10016091 * (256/254))`), then it's rounded up to the next nearest power-of-two value as "pieces" stored in Filecoin have to fit neatly into a proving tree (`2**Math.ceil(Math.log2(Math.ceil(10016091*(256/254))))`).
+  * The additional padding is done transparently, but it leads to wonky numbers in the deal process that aren't explained well (and online documentation is inconsistent on this and doesn't explain it well - TODO some PRs)
+  * The most consistent number shown is power-of-two _minus_ the fr32 padding. (`2**Math.ceil(Math.log2(Math.ceil(10016091*(256/254))))/(256/254)/1024/1024` as `MiB`, or `15.875`, presented as `15.88 MiB` by Lotus - clear as mud).
+  * Aside: an implication of padding is that when packing data in for making efficient deals, you should get as close to the "power-of-two - fr32 padding" size as possible. The larger the data, the more waste you'll have. We had to do this for Filecoin Discover data to edge up as close as possible to 1 GiB pieces.
```

Comment:
The inverse of this implication is that certain miner/client combos with access to datacap make deals that pad to precisely 16.001GiB, to enjoy a full 10x sector but effectively "lock" just half the data they have available.

Reply:
this is back to my question above - "deals that pad to precisely 16.001GiB" - I was under the impression that "deal size" that you pay for is the entire thing, but apparently it's not? you only pay for pre-padded data?

Reply:
No, you are supposed to pay for the entire thing. Transfer savings is what makes this attractive.

Reply:
got it, that's really interesting context, I thought that the padded content gets transferred but if not then that's a nice efficiency gain on data transfer

Reply:
The only thing transferred is the indiviual blocks of the DAG, selector-style. `car` files themselves never come into the equation, let alone the padding on top of them.
Remember this entire thing being a... thing: https://github.com/filecoin-project/go-fil-markets/issues/365
Storage is "mandatory-graphsync" in exactly the same way ( currently ).

Reply:
yeah, I do remember now, all the gory graphsync stuff comes into focus in this point doesn't it? makes sense but is absolutely obscured by Lotus always talking about this "datasize" thing